### PR TITLE
Write nomination errors to assets file

### LINKED ISSUE #7717: Nomination errors, such as bad data (version, tfm etc), are not propagated to Visual Studio.
Take this project; 

```
<Project>
  <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net472</TargetFrameworks>
  </PropertyGroup>
  
  <ItemGroup>
    <PackageReference Include="NuGet.Common" Version="blabla" />
  </ItemGroup>


  <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
</Project>
```

Restore from the commandline. 
The output contains something like the below:

```
Microsoft (R) Build Engine version 16.0.360-preview+g9781d96883 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 1/22/2019 2:23:03 PM.
Project "C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp51\ConsoleApp51\ConsoleApp51.csproj" on node 1 (Restore target(s)).
C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(119,5): error : 'blabla' is not a valid version string. [C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp51\ConsoleApp51\ConsoleApp51.csproj]
Done Building Project "C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp51\ConsoleApp51\ConsoleApp51.csproj" (Restore target(s)) -- FAILED.


Build FAILED.

"C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp51\ConsoleApp51\ConsoleApp51.csproj" (Restore target) (1) ->
(Restore target) ->
  C:\Program Files (x86)\Microsoft Visual Studio\2019\IntPreview\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(119,5): error : 'blabla' is not a valid version string. [C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp51\ConsoleApp51\ConsoleApp51.csproj]

    0 Warning(s)
    1 Error(s)
```

Restore in VS. 

Nothing is seen in the error list. 

The output is something like: 
```
All packages are already installed and there is nothing to restore.
Time Elapsed: 00:00:00.1668614
========== Finished ==========
```

What's happening here is that the restore is being skipped because the package spec could not be generated properly. 

//cc @rrelyea 
This will become even more apparent when we add the exact version range checks for PackageDownload

### PR DESCRIPTION
**Pull request by zivkan**
## Bug

Fixes: NuGet/Home#7717
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: Pass nomination errors through to restore, to write errors to the assets file, so the project system can display them in Visual Studio's error list.

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:

### CODE REVIEW

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/IProjectSystemCache.cs
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/IProjectSystemCache.cs:113]: This is a public interface, what is the policy of changing the interfaces?

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectSystemCache.cs
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectSystemCache.cs:273]: Public method - consider an overload?

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/IProjectSystemCache.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/IProjectSystemCache.cs:113]: The project is `NuGet.VisualStudio.Common`. Like the other `src\NuGet.Client\*` projects, we don't publish this as a package, meaning no-one can have an API or ABI dependence on this project. API breaking changes are only important for `src\NuGet.Core\*` projects, and our `IVs*` interfaces.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectSystemCache.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/ProjectSystemCache.cs:273]: As the other comment, this is not an assembly that we provide in a package.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: Will be the nomination errors added to the assets file?

Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: yes, the exception is added to the message list on lines 131 to 134, and on line 144 it gets added to the project system cache, which is then retrieved in NuGet.Commands's RestoreCommand.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs
Comment (cristinamanum) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs]: Internal for testing?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (cristinamanum) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:241]: If _success was false from the beginning why it was needed to do the lock and noop evaluation and not fail directly?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs
Comment (cristinamanum) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs:188]: Why Additional in the name? There are not any messages yet.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/Resources.resx
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/Resources.resx]: The error can be not because the project information failed to be read but maybe evaluated.  Maybe around the line "A failure was encountered while evaluating the project information: {0}" ?

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (cristinamanum) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: NU1010 error will be a wrapper error for anything falling here (invalid TFM, etc).  Maybe it would be worth to have a NuGet exception that will allow different project data inspectors/evaluators to send their own NU**** messages ?

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (cristinamanum) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs]: additioalMessages -> additionalMessages

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:110]: Should this one be deprecated? Or remove? I understand it's part of NuGet.Packagemanagement so that might be the thing giving you a pause but technically this is break too :)

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: Why are we throwing?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs]: I don't think the spec validation should be failing?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:358]: How can we have error messages in addition to the ones provided from the parsing?

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs]: Why not just re-use https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1105/ ? 

They communicate the same thing.

File: src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs]: Not a fan of the factory methods in such a low level class :(
There's no use beyond what you are doing is there? 
I don't  think we enough use-cases to define a good API. 
Why can't it be private/internal to the assembly that creates them.

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs]: same as the above.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:31]: I'm adding it here, but it's a general question about the restore execution.
It's a bit hard to understand all the scenarios given where the additional check is happening. 

Can you explain the following scenario and ensure that it works as expected. 

Say I have 2 package references and both of them have a bad version.
Because we probably fail fast (do we?), the error message talks about the first version. 
Now I fix that one, nomination happens again, and I get another error. is the error message updated? 

As is, I'm afraid that it might no-op and give the old warning message.

**Review by nkolev92 — COMMENTED** [review]
Looks great overall. 

A couple of questions/clarifications needed.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:110]: No, I think it's fine to have both. At this time, only VS nominations have the possibility of having restore messages. I don't know if in the future we'll want to make command line restore create assets files when the project fails to evaluate. At that time we could consider if it should become deprecated.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/Resources.resx
Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/Resources.resx]: Only people with a reasonable MSBuild knowledge will understand what an evaluation is. To most .NET developers, particularly beginners, the more specific message may be difficult to understand.

What do you think about this?

```suggestion
    <value>Unable to parse the project information: {0}</value>
```

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: I want nomination to be resilient to unknown and future errors. There are multiple improvements we can make to this. For example, more specific NU codes for specific things like invalid project TFM or invalid version on a package ref. Another example is right now the error is thrown as soon as the first error is found. So if there are multiple problems, only the first is reported. Ideally that should be improved as well, but it can be done in another PR.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:110]: These APIs are part of NuGet.PackageManagement, and as such they are not part of msbuild/dotnet restore.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: `_logger` just goes to the VS activity log, but we have no way of knowing how often customers are actually getting this as it's otherwise unreported. Once this PR is merged, the only known way of reaching this exception handler will no longer be possible, so I thought it'd be reasonable to allow it to generate Watson reports again.

Do you have another suggestion? Should I make it generate a telemetry event?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs]: [It validates project TFMs](https://github.com/NuGet/NuGet.Client/blob/3803820961f4d61c06d07b179dab1d0439ec0d91/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/SpecValidationUtility.cs#L131-L162). When the TFM in the project file is invalid, I didn't want to hardcode a TFM just to make the validation pass, as that'll get written to the assets file and I didn't want the assets file to list a TFM that wasn't specified in the project file.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:241]: Initially I was hoping no-op would still work, so we could avoid having to write the assets file more than once, until the nomination error was fixed. Since I couldn't get that to work, I guess it's pointless.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRequest.cs:188]: Very true. I was thinking they're additional to whatever messages restore generates.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:358]: We can't but since these things happen at different levels of abstraction, we can't guarantee it. The "additional messages" passed in the restore args could only be warnings (if we change something in the future, or someone uses our packages and does it in their own code).

File: src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Common/Errors/NuGetLogCode.cs]: Because I didn't spend enough time looking up existing error codes. Thanks for letting me know about it.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: I don't think we should have Watson reports for users errors. 
At least what I thought the idea was a while back when we stopped throwing.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: As far as the activity log goes, it's only for us...customers shouldn't see it, the customers will see the issues in the error window.

As far as a telemetry event goes, I'll leave that at your discretion. Don't know if we have a guidance for these things, beyond fire and forget events.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/DependencyGraphSpecRequestProvider.cs]: :( 
Can we leave the validation for the other projects? Is that a consideration?
Maybe in a future PR, don't think it's worth blocking this PR on that.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: I think adding more granular errors will not have a great ROI at this point in time.
These bad input situations are very rare. We should only invest once we know the benefit is there. 

I imagine it'd be super painful to rework it.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:31]: With the current implementation, no-op is skipped (see the change to `EvaluateCacheFile`), so as long as there are messages in the `RestoreArgs`, it will never no-op. This lets the error message(s) be written to the assets file when the message(s) change.

You're right about it only showing one error at a time at the moment. That's an improvement that can be made in `NominateProjectAsync`, and why the `AdditionalMessages` property in `RestoreArgs` is a list, not a single item.

During development, I had the problem that you described, when the error changed, the assets file wasn't being updated. But it was since fixed, unfortunately by disabling no-op.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:31]: It's fine if no-op is skipped, it's just a bit difficult to follow. 
Maybe a refactoring is required in that part of the code.
Feel like with noop + lock file + CPVM scenarios things are declared all over the place which made it difficult to read.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:241]: I looked into this further, and there are several variables that are needed at the end of the method, during asset file and DGSpec writing. I don't think that it would be worthwhile to refactor the code to skip the no-op check, but still get the values for these variables.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/Resources.resx
Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/Resources.resx]: Resolving, as I'm going to reuse NU1105 instead of creating a new code.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: Indeed we shouldn't have Watson reports for user errors, but the new nomination error handling should catch user errors gracefully. If an exception somehow gets to this comment's catch block, that means that graceful nomination error handling didn't work, and the user error will not be reported in the assets file.

In other words, there was an application error in being unable to handle the root cause gracefully.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: I'd rather us handle it than rely on some other teams' infrastructure to report this.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: @zivkan before I review again, can you look at the above, specifically 

> I'd rather us handle it than rely on some other teams' infrastructure to report this.

If we think it's important enough, we should track is ourselves rather than rely on partner reporting.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: For CPVM, we need to fail restore with specific NUxxx error code. One way that I was thinking to achieve this is to do the validation and throw the exception with the NUxxx error code. This is only one example, but probably there are other cases with the same pattern.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: > For CPVM, we need to fail restore with specific NUxxx error code.

Why? It's just like any input error.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs]: @nkolev92 updated to report errors without creating watson crashes.

**Comment by nkolev92** [comment]
@zivkan you have compile errors :D

I'll review this today.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs:109]: nit: no need to remove this line.

File: src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/TelemetryUtility.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/Telemetry/TelemetryUtility.cs]: I'd consider unwrapping the exception to get the inner most one.
Maybe report both?

**Review by nkolev92 — APPROVED** [review]
Looks great. 

I think you need to unwrap the posted exception but don't want to block this PR on it. 

You can either do it in the same PR or a future one.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs
Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: @cristinamanum I'm resolving this comment, but I'm a bit concerned why you think CPVM should duplicate all existing errors and give them new codes. Perhaps we should talk about that.

Comment (cristinamanum) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: I was not saying CPVM should duplicate all the existing errors and give them new error codes. I was referring to the fact that this CPVM error will be added with NU error code based on a  suggestion 
"_I think that this error needs an NU code, so if customers see it, then can easily find a docs page to help explain the issue._" [here](https://github.com/NuGet/NuGet.Client/pull/3198#discussion_r369864252 ) to have a NU** code. 

If we decide that any input error we treat it as same NU error code is perfectly fine. However if we like to have the flexibility to differentiate the user inputs in different NU error codes the current design approach will not help. 

The current PR can go ahead and be merged as the above topic will probably need more discussions,

Comment (zivkan) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs:127]: edit: clarified something I meant.

Ok, there was a misunderstanding. Since CPVM downgrade errors are using a different NU code to non-CPVM downgrade warnings (😢 but at least it's more understandable because of errors vs warnings), I mis-interpreted "*For CPVM, we need to fail restore with specific NUxxx error code*" and thought you wanted all NU codes duplicated to be different in cpvm and non-cpvm projects.

Instead, I now understand you just mean that when errors occur, you want the specific error to be reported, not NU1010 or NU1105 for all different reasons. And CPVM has some warnings and errors that do not exist in non-CPVM cases. As I mentioned in my first reply to your comment on this line of code, this is an improvement for another PR. I agree with Nikolche that there is insufficient benefit at this time.

Also, in case it's not clear, this error handling (reporting NU1010/NU1105) is only run when converting `IVsProjectRestoreInfo` from project system nomination into a DGSpec to put into the project cache. Errors that get reported during `RestoreCommand` will retain their NU code, and it will also work from both command line restores and VS restores. Validation that gets done during nomination needs to be duplicated in command line DGSpec generation, so it's really beneficial to do it in the `RestoreCommand` instead.

**Pull request merged**
